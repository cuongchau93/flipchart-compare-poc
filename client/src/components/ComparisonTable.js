import React, { Component } from 'react';
import { Table } from 'reactstrap';
import FeatureRow from './FeatureRow';
import '../styles/ComparisonTable.css';

class ComparisonTable extends Component {
    render() {
        const rows = [];

        if(this.props.features){
            if(this.props.features.Highlights){
                Object.keys(this.props.features.Highlights).forEach((spec)=>{
                    rows.push(
                        <FeatureRow
                            className="highlighted"
                            name={spec}
                            feature1={this.props.features.Highlights[spec][0]}
                            feature2={this.props.features.Highlights[spec][1]}
                            key={spec+this.props.features.Highlights[spec][0]+this.props.features.Highlights[spec][1]} />
                    );
                });
            }

            if(this.props.features.Specifications){
                Object.keys(this.props.features.Specifications).forEach((spec)=>{
                    rows.push(
                        <FeatureRow
                            className="normal"
                            name={spec}
                            feature1={this.props.features.Specifications[spec][0]}
                            feature2={this.props.features.Specifications[spec][1]}
                            key={spec+this.props.features.Specifications[spec][0]+this.props.features.Specifications[spec][1]} />
                    );
                });
            }
        }

        return (
            <Table responsive hover bordered>
                <thead>
                    <tr>
                        <th>Features</th>
                        <th>Product 1</th>
                        <th>Product 2</th>
                    </tr>
                </thead>
                <tbody>{rows}</tbody>
            </Table>
        );
    }
}

export default ComparisonTable;