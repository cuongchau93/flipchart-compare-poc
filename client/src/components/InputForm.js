import React, { Component } from 'react';
import '../styles/InputForm.css';
import { Button, Form, FormGroup, Label, Input} from 'reactstrap';

class InputForm extends Component {
    constructor(props) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
        this.state = {
            loading: false
        }
    }

    handleSubmit(event) {
        if(event.target.url1.value && event.target.url2.value){
            this.props.onSubmission(event.target.url1.value ,event.target.url2.value);
        }
        event.preventDefault();
    }

    render() {
        return (
            <Form onSubmit={this.handleSubmit} id="myForm">
                <FormGroup>
                    <Label for="url1">First URL</Label>
                    <Input type="url" id="url1" name="url1" defaultValue="https://www.lazada.sg/products/samsung-galaxy-s8-64gb-midnight-black-i106217842-s108237124.html" />
                </FormGroup>
                <FormGroup>
                    <Label for="url2">Second URL</Label>
                    <Input type="url" id="url2" name="url2" defaultValue="https://www.lazada.sg/apple-iphone-8-256gb-2gb-ram-grey-60291398.html" />
                </FormGroup>
                <Button>Submit</Button>
            </Form>
        );
    }
}
export default InputForm;