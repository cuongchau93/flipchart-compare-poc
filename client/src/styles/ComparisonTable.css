thead {
  background-color: #b9c9fe; }

.highlighted {
  background-color: #e8edff; }
