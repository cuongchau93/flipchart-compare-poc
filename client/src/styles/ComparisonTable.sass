$lightPurple: #b9c9fe
$lighterPurple: rgb(232, 237, 255)

thead
    background-color: $lightPurple

.highlighted
    background-color: $lighterPurple
