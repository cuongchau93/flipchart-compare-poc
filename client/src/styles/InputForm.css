#myForm {
  background-color: lightgray;
  padding: 20px; }

#myForm input {
  border: 1px solid lightgray;
  padding: 12px 20px;
  border-radius: 4px;
  box-sizing: border-box; }
