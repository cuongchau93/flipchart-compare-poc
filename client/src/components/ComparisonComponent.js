import React, { Component } from 'react';
import InputForm from './InputForm';
import ComparisonTable from './ComparisonTable';
import LoadingSpinner from './LoadingSpinner';

class ComparisonComponent extends Component {

    constructor(props){
        super(props);

        this.state = {
            features: {},
            loading: false,
            error: ""
        };
        this.handleSubmission = this.handleSubmission.bind(this);
    }

    callApi = async (u1,u2) => {
        const response = await fetch('/lazada/services/compare', {
            method: 'POST',
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                url1: u1,
                url2: u2,
            }),
        });

        const body = await response.json();
        if (response.status !== 200)
            throw Error(body.message);

        return body;
    };

    handleSubmission(u1, u2){
        this.setState({
            loading: true
        });

        this.callApi(u1,u2).then( response => {
            this.setState({
                features: response,
                loading: false
            });

        }).catch(err =>{
            this.setState({
                error: err.message,
                loading: false
            });
        });
    };

    render() {
        return (
            <div>
                {this.state.error !== "" &&
                <div className="alert alert-danger" onClick={() => {
                    this.setState({error: ""});
                }}>{this.state.error}</div>
                }

                <InputForm onSubmission={this.handleSubmission}/>

                <br/>
                {this.state.loading && <LoadingSpinner />}

                {Object.keys(this.state.features).length !== 0 &&
                <ComparisonTable features={this.state.features} />}

            </div>
        );
    }
}

export default ComparisonComponent;