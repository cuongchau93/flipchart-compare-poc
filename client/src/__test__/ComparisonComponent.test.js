import React from 'react';
import ComparisonComponent from '../components/ComparisonComponent';
import InputForm from '../components/InputForm';
import { mount, shallow } from 'enzyme'
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ComparisonTable from '../components/ComparisonTable';

configure({ adapter: new Adapter() });

describe("ComparisonComponent " , () => {
    const component = shallow(
        <ComparisonComponent />
    );

    it('renders and matches our snapshot', () => {
        const component = renderer.create(
            <ComparisonComponent />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contains an InputForm subcomponent with no table', () => {
        expect(component.find(InputForm)).toHaveLength(1);
        expect(component.find(ComparisonTable)).toHaveLength(0);
    });
});