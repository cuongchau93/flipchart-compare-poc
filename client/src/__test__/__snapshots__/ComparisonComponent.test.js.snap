// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComparisonComponent  renders and matches our snapshot 1`] = `
<div>
  <form
    className=""
    id="myForm"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className=""
        htmlFor="url1"
      >
        First URL
      </label>
      <input
        className="form-control"
        defaultValue="https://www.lazada.sg/products/samsung-galaxy-s8-64gb-midnight-black-i106217842-s108237124.html"
        id="url1"
        name="url1"
        type="url"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        className=""
        htmlFor="url2"
      >
        Second URL
      </label>
      <input
        className="form-control"
        defaultValue="https://www.lazada.sg/apple-iphone-8-256gb-2gb-ram-grey-60291398.html"
        id="url2"
        name="url2"
        type="url"
      />
    </div>
    <button
      className="btn btn-secondary"
      onClick={[Function]}
      type={undefined}
    >
      Submit
    </button>
  </form>
  <br />
</div>
`;
